src/main/java/org/terasology/Lakes/LakeRasterizer.java

```diff
@@ -23,74 +23,29 @@
-        LavaFacet lavaFacet = chunkRegion.getFacet(LavaFacet.class);
-        SurfaceHeightFacet surfaceHeightFacet = chunkRegion.getFacet(SurfaceHeightFacet.class);
-
-        for (Vector3i position : chunkRegion.getRegion()) {
-
-            Lake lake = lakeFacet.getNearestLake(position);
-            Lake lavaLake = lavaFacet.getNearestLake(position);
-
-            if (lake.isNotNull() && lake.isInRange(position)) {
-
-                float surfaceHeight = surfaceHeightFacet.getWorld(position.x(), position.z());
-                float lakeDepth = lakeDepthFacet.getWorld(position.x(), position.z());
-                float lakeHeight = lakeHeightFacet.getWorld(position.x(), position.z());
-
-                if (lake.LakeContains(position) && position.y() <= lake.getWaterHeight() && (position.y() >= lake.getWaterHeight() - lakeDepth ||
-                        position.y() > surfaceHeight)) {
-                    chunk.setBlock(ChunkMath.calcRelativeBlockPos(position), water);
-                } else if (lake.OuterContains(position) && position.y() <= lake.getWaterHeight() && position.y() >= surfaceHeight) {
-                    chunk.setBlock(ChunkMath.calcRelativeBlockPos(position), sand);
-                } else if (lake.LakeContains(position) && position.y() > lake.getWaterHeight() && position.y() <= lake.getWaterHeight() + lakeHeight) {
-                    chunk.setBlock(ChunkMath.calcRelativeBlockPos(position), air);
-                }
-            }
-
-            if (lavaFacet.isEnabled() && lavaLake.isNotNull() && lake.isInRange(position)) {
 
-                float surfaceHeight = surfaceHeightFacet.getWorld(position.x(), position.z());
-                float lakeDepth = lakeDepthFacet.getWorld(position.x(), position.z());
-                float lakeHeight = lakeHeightFacet.getWorld(position.x(), position.z());
-
-                if (lavaLake.LakeContains(position) && position.y() <= lavaLake.getWaterHeight() && (position.y() >= lavaLake.getWaterHeight() - lakeDepth ||
-                        position.y() > surfaceHeight)) {
-                    chunk.setBlock(ChunkMath.calcRelativeBlockPos(position), lava);
-                } else if (lavaLake.OuterContains(position) && position.y() <= lavaLake.getWaterHeight() && position.y() >= surfaceHeight) {
-                    chunk.setBlock(ChunkMath.calcRelativeBlockPos(position), stone);
-                } else if (lavaLake.LakeContains(position) && position.y() > lavaLake.getWaterHeight() && position.y() <= lavaLake.getWaterHeight() + lakeHeight) {
-
-                    chunk.setBlock(ChunkMath.calcRelativeBlockPos(position), air);
+        for (Lake lake : lakeFacet.getLakes()) {
+            for (Vector3i pos : lake) {
+                if (chunkRegion.getRegion().encompasses(pos)) {
+                    Block block = pos.y > lake.surfaceHeight ? air : lake.liquid;
+                    chunk.setBlock(ChunkMath.calcRelativeBlockPos(pos), block);
```

Comment:
This is neat! :nerd_face:

Reply:
It allows somewhat less explicit control over how intersecting lakes are dealt with than the previous approach, but it should also be much more efficient.